src/vmm/src/builder.rs

```diff
@@ -471,37 +470,60 @@ pub fn build_microvm_from_snapshot(
         vcpu_count,
     )?;
 
-    #[cfg(target_arch = "x86_64")]
-    // Check if we need to scale the TSC.
-    // We start by checking if the CPU model in the snapshot is
-    // the same as this host's. If they are the same, we don't
-    // need to do anything else.
-    if !is_same_model(&microvm_state.vcpu_states[0].cpuid) {
-        // Extract the TSC freq from the state.
-        // No TSC freq in snapshot means we have to fail-fast.
-        let state_tsc = microvm_state.vcpu_states[0]
-            .tsc_khz
-            .ok_or_else(|| {
-                MicrovmStateError::IncompatibleState(
-                    "Error configuring the TSC, frequency not present in snapshot.".to_string(),
-                )
-            })
-            .map_err(RestoreMicrovmState)?;
-
-        // Scale the TSC frequency for all VCPUs, if needed.
-        if vcpus[0]
-            .kvm_vcpu
-            .is_tsc_scaling_required(state_tsc)
-            .map_err(|err| MicrovmStateError::IncompatibleState(err.to_string()))
-            .map_err(RestoreMicrovmState)?
-        {
-            for vcpu in &vcpus {
-                vcpu.kvm_vcpu
-                    .set_tsc_khz(state_tsc)
-                    .map_err(|err| MicrovmStateError::IncompatibleState(err.to_string()))
-                    .map_err(RestoreMicrovmState)?;
+    // CPUID is supported when:
+    // - We are on an x86 archtecture with `see` enabled and `sgx disabled`.
+    // - We are on an x86_64 architecture with `sgx` disabled
+    #[cfg(any(
+        all(target_arch = "x86", target_feature = "sse", not(target_env = "sgx")),
+        all(target_arch = "x86_64", not(target_env = "sgx"))
+    ))]
+    {
+        // Convert CPUID to better format
+        let raw_snapshot_cpuid = cpuid::RawCpuid::from(microvm_state.vcpu_states[0].cpuid.clone());
+        let snapshot_cpuid = cpuid::Cpuid::try_from(raw_snapshot_cpuid).expect("TODO Remove this");
+        // This is safe as we check for CPUID support.
+        let host_cpuid = unsafe { cpuid::Cpuid::new().unwrap() };
```

Comment:
Since KVM is doing the emulation, here we care about the CPUID that KVM supports, not the actual host CPUID. KVM will expose the supported CPUID as a subset of the host. (see https://elixir.bootlin.com/linux/v5.10.136/source/arch/x86/kvm/cpuid.c#L347 and https://elixir.bootlin.com/linux/v5.10.136/source/arch/x86/kvm/vmx/vmx.c#L7400)

So `host_cpuid` should be constructed from the result of the KVM_GET_SUPPORTED_CPUID ioctl.

Reply:
Fixed it.

Reply:
Yes that's it.